### PR TITLE
[Issue 834] label formatters

### LINKED ISSUE #834: [FEAT][CV] Implement label formatters for common formats
Label Formatter can accept a callable, or a string signifying that the labels come in a common format. The built in converter functions for these string values need implementing.

### PR DESCRIPTION
**Pull request by yromanyshyn**
resolves #834

### CODE REVIEW

File: deepchecks/vision/utils/detection_formatters.py [hunk 1]
```diff
@@ -203,3 +207,176 @@ def validate_prediction(self, batch_predictions, n_classes: int, eps: float = 1e
         if batch_predictions[0].shape[1] != 6:
             raise DeepchecksValueError('Check requires detection predictions to be a list of 2D tensors, when '
                                        'each row has 6 columns: [x, y, width, height, class_probability, class_id]')
+
+
+def verify_bbox_format_notation(notation: str) -> List[str]:
```
Comment (noamzbr) [inline deepchecks/vision/utils/detection_formatters.py:212]: We're missing one important format option here (sorry, my bad for forgetting to detail it in the issue) - we need an "n" option, which means that all coordinates are normalized rel to image size, rather than the default "absolute" coordinates which are in pixels.

Comment (yromanyshyn) [reply deepchecks/vision/utils/detection_formatters.py:212]: ok, I will add

File: deepchecks/vision/dataset.py [hunk 2]
```diff
@@ -98,17 +98,6 @@ def __init__(self,
         self._data = data_loader
         self.label_transformer = label_transformer
         self.image_transformer = image_transformer or ImageFormatter(lambda x: x)
-
```
Comment (noamzbr) [inline deepchecks/vision/dataset.py:101]: why delete this?

**Review by noamzbr — COMMENTED** [review]
Main comment - need to complete the following section of formatter_objects.rst:

Converting From Common Formats
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
TBD (Not implemented yet)

File: deepchecks/vision/dataset.py (hunk 2 above)
Comment (yromanyshyn) [reply deepchecks/vision/dataset.py:101]: code duplicates. Probably someone merged incorrectly

File: deepchecks/vision/utils/detection_formatters.py (hunk 1 above)
Comment (yromanyshyn) [reply deepchecks/vision/utils/detection_formatters.py]: added

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 3]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+To initialize `DetectionLabelFormatter` for common label formats the notion of the bbox format notation can be used:
+
+.. code-block:: python
+
+    import torch
+    from deepchecks.vision.utils import DetectionLabelFormatter
+
+    yolo_formatter = DetectionLabelFormatter('lcxcywhn') 
+    pascal_formatter = DetectionLabelFormatter('lxyxy')
+    coco_formatter = DetectionLabelFormatter('lxywh')
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:89]:
```suggestion
    coco_formatter = DetectionLabelFormatter()  # This is the default format of 'lxywh'
```

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 4]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+    albumentations_formatter = DetectionLabelFormatter('lxyxyn')
+
+By using provided bbox format notation deepchecks will know how to transform labels to the format required by deepchecks, which is 'lxywh'.
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:92]:
```suggestion
Based on the outlined bbox format notation deepchecks will transformed the labels to the format required by deepchecks, which is 'lxywh'.
```

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 5]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+
+Elements of the bbox format notation:
+    * l - label (class) of the object;
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:95]:
```suggestion
    * l - label (class index) of the object;
```

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 6]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+    * xy - top-left or bottom-right corner of the bbox. (First 'xy' element is always interpreted as x-min, y-min, and second as x-max, y-max). Coordinates of the bounding box's corners are represented with respect to the top-left corner of the image which has (x, y) coordinates (0, 0);
+    * cxcy - center of the bbox;
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:97]:
```suggestion
    * cxcy - xy measured from the center of the image, meaning the center of the image is (0, 0);
```

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 7]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+    * wh - width and height of the bbox;
+    * n - element placed at the beggining or at the end of the notation indicates whether coordinates are normalized or not. Which means that coordinates of the bbox are representated not in pixels but as ration of x / image-width and y / image-height.
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:99]:
```suggestion
    * n - placed at the beginning or at the end of the notation to indicate whether the coordinates are normalized or not. Normalized coordinates are not measured in pixels, but as rather ratio of x / image-width and y / image-height.
```

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 8]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+
+An important moment to note is that `DetectionLabelFormatter` do not work with labels represented by JSON, XML, or python dictionaries.
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:101]:
```suggestion
It is important to note that `DetectionLabelFormatter` does not work with labels represented by JSON, XML, or python dictionaries.
```

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 9]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+The acceptable objects types for `DetectionLabelFormatter` are numpy arrays, pytorch tensors, and python sequence-like objects.
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:102]:
```suggestion
The label batch processed by `DetectionLabelFormatter` must be an iterable of numpy arrays.
```
Does the code really work on label batches which are iterables of tensors or other stuff?

File: docs/source/user-guide/vision/formatter_objects.rst [hunk 10]
```diff
@@ -73,9 +73,75 @@ For object detection tasks, it is required to implement the following formatters
...
+
+Example to represent the thought:
```
Comment (noamzbr) [inline docs/source/user-guide/vision/formatter_objects.rst:104]:
```suggestion
Example of using the notation format to get a DetectionLabelFormatter converting from yolo format to deepchecks format:
```

File: docs/source/user-guide/vision/formatter_objects.rst (hunk 9 above)
Comment (yromanyshyn) [reply docs/source/user-guide/vision/formatter_objects.rst:102]: > Does the code really work on label batches which are iterables of tensors or other stuff?

do not know what you mean by `other stuff` but yes it works with pytorch tensors
